Integrate crate Url + routing refactor

MartinKavik opened the pull request:
@David-OConnor 
I tried to refactor routing a little bit. 
I don't want to merge this PR because it uses crate `Url` in the one of methods and this library increases WASM file size too much (https://github.com/servo/rust-url/issues/557). 
PR (almost) doesn't change logic, only code style, so you can pick your "favorite" parts and implement them in your original PR. Only critical change which should be moved into original PR is `#[test]` -> `#[wasm_bindgen_test]` because these tests are ignored at the moment.
PR is quite short, ask questions if something doesn't make sense. Thanks.

(OT: It seems that it's not possible to set PR as Draft, once created)

AlterionX commented on src/routing.rs:
```diff
@@ -372,9 +343,13 @@ where
+
     use super::*;
 
-    #[test]
```
Is it possible to leave this annotation in as well so that both `test` and `wasm_bindgen_test` are available? Just curious.

MartinKavik replied:
With both:
- `cargo make test_h_firefox` seems to work.
- `cargo test` seems to run test successfully, then fails on some errors

So it seems to work at least for tests which doesn't need browser to run.

David-OConnor commented:
As you noticed from a comment about Regex in the original PR, it's unfortunate we have to handwrite code that would be better suited to a dependency, especially given Rust's philosophy of keeping the standard library small.

Re the `test` tag: It's too bad we can't just use normal tags for code that doesn't need wasm.

David-OConnor commented:
I like your code integrating `Url`'s built-in parsing; would love to replace the current code with this if the WASM-size increase can be minimized. I fixed the tests per your changes on the main PR.

MartinKavik commented:
> I fixed the tests per your changes on the main PR.

Just note - I've changed also these two things:
1. `"/path/#hash?sea=rch"` -> `"/path?sea=rch#hash"`
   - Everything after hash is ignored from the server point of view (it only lives in browser), so crate `rust-url` returns `search` == `None` in your original case, because `search`'s main purpose is to send parameters to backend. 
1. `"/path/#hash"` -> `"/path#hash"` (trailing slash in more cases)
   - Crate `rust-url` returns also trailing empty string in path segments so you can differentiate between files and directories. It's the second difference between Seed's behavior and that library.

MartinKavik commented:
I'm closing it because of https://github.com/David-OConnor/seed/issues/226#issuecomment-544923393. We can reopen it / create a new PR once we have more lightweight URL library.
